build_runner/lib/src/package_graph/target_graph.dart

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:build_config/build_config.dart';
+
+import 'package_graph.dart';
+
+/// Like a [PackageGraph] but packages are further broken down into modules
+/// based on build config.
+class TargetGraph {
+  /// All [TargetNode]s indexed by `"$packageName:$targetName"`.
+  final Map<String, TargetNode> allModules;
+
+  TargetGraph._(this.allModules);
+
+  static Future<TargetGraph> forPackageGraph(PackageGraph packageGraph,
+      {Map<String, BuildConfig> overrideBuildConfig}) async {
+    overrideBuildConfig ??= const {};
+    final modules = <String, TargetNode>{};
+    for (final package in packageGraph.allPackages.values) {
```

Comment:
nit: consider `Future.wait` to speed this up a bit

Reply:
Will look at optimizations in a followup.